File: src-tauri/src/quit.rs
// The quit guardrail.
//
// Quitting Operator ends every lane's pty, and with it whatever turn each agent was in the
// middle of. Until now nothing asked: the red traffic-light button destroyed the window, the
// window store emptied, and the app exited — which is exactly how the 2026-08-14 accident ran
// (a stray image drop navigated the webview to a file:// URL, the user closed what no longer
// looked like Operator, and every lane went with it).
//
// WHAT THIS DOES AND DOES NOT CATCH. Three quit paths remain unguarded, by decision, and this
// module does not pretend otherwise:
//
//   Dock (right-click) → Quit   native `-[NSApplication terminate:]`; reaching it needs an
//                               unsafe `applicationShouldTerminate:` override on the shared
//                               NSApplicationDelegate, outside tao 0.35.3's public surface.
//   macOS logout / restart      same delegate method. A guard that vetoes a system shutdown is
//                               a bug, so this one is not merely unbuilt — it is refused.
//   Force Quit                  SIGKILL. Impossible by definition.
//
// The app menu's ⌘Q *is* caught, but only because `build_menu` below replaces the predefined
// Quit item: `PredefinedMenuItem::quit` binds the native `sel!(terminate:)` selector
// (muda-0.19.2/src/platform_impl/macos/mod.rs:994) and never reaches Tauri's MenuEvent or
// RunEvent at all. A custom item carrying the same ⌘Q accelerator routes through a normal
// MenuEvent instead. The tray's `MenuBuilder::quit()` was the same predefined item and is
// replaced the same way.

use std::sync::atomic::{AtomicBool, AtomicU64, Ordering};
use std::time::Duration;

use serde::Serialize;
use tauri::menu::{Menu, MenuItem, MenuItemKind};
use tauri::{AppHandle, Emitter, Manager, State};

use crate::transcript::{LiveLane, LiveLanes};

/// App-menu item ids. Namespaced so they can never collide with the tray's own ids.
pub const MENU_QUIT: &str = "operator:quit";
pub const MENU_QUIT_NOW: &str = "operator:quit-now";
/// Tray menu item id — the tray listens on its own handler, see `build_tray`.
pub const TRAY_QUIT: &str = "quit";

/// How long the webview gets to mount the dialog and ack before we fall back to a native ask.
/// The fallback is not a nicety: it is the whole reason the count lives in Rust. When the
/// renderer is navigated away, frozen, or mid-respawn there is no React app to render anything.
const ACK_MS: u64 = 400;

/// The phases that mean "mid-turn". `waiting` is included deliberately: an agent blocked on YOU
/// is the precise lane you forgot about. A live-but-idle lane does not trigger the dialog — it
/// would then fire on nearly every quit, which is how a guard trains its own dismissal.
fn is_busy(phase: &str) -> bool {
    matches!(phase, "running" | "compacting" | "waiting")
}

pub struct QuitGuard {
    /// The user has answered "quit" (or bypassed): every hook becomes a pass-through.
    confirmed: AtomicBool,
    /// A prompt is on screen. ⌘Q pressed twice must not quit by repetition.
    prompting: AtomicBool,
    /// The webview acked that it mounted the dialog.
    acked: AtomicBool,
    /// Mirrors the "Ask before quitting with agents running" preference (default on). The
    /// preference itself lives in the renderer's localStorage with the app's other switches;
    /// the renderer mirrors it here on mount and on every toggle. If it never gets the chance,
    /// this stays ON — the direction a default should fail in.
    ask: AtomicBool,
    /// Incremented per prompt, so a late ack or a late native-dialog answer from a prompt the
    /// user already resolved cannot answer the next one.
    round: AtomicU64,
}

impl QuitGuard {
    /// The user has answered "quit" — the ExitRequested hook must let that exit through.
    pub fn is_confirmed(&self) -> bool {
        self.confirmed.load(Ordering::SeqCst)
    }
}

impl Default for QuitGuard {
    fn default() -> Self {
        QuitGuard {
            confirmed: AtomicBool::new(false),
            prompting: AtomicBool::new(false),
            acked: AtomicBool::new(false),
            ask: AtomicBool::new(true),
            round: AtomicU64::new(0),
        }
    }
}

/// What the dialog renders from. The frontend enriches each lane by `terminalId` (lane name,
/// accent, the chat-signal wording) and falls back to what is here for any it cannot match —
/// so there is no loading state and no store read, and a webview with a stale session list
/// still names the right lanes.
#[derive(Clone, Serialize)]
#[serde(rename_all = "camelCase")]
struct QuitRequest {
    /// Busy lanes only, unordered — the frontend owns presentation order.
    lanes: Vec<LiveLane>,
    /// Live lanes that are NOT busy. They will be ended too, and are counted in one line.
    idle: usize,
}

fn raise_main(app: &AppHandle) {
    // Tray and menu-bar paths can fire while the window is hidden or minimised. Without this
    // the app looks wedged: an invisible dialog holding a quit nobody can see.
    if let Some(w) = app.get_webview_window("main") {
        let _ = w.unminimize();
        let _ = w.show();
        let _ = w.set_focus();
    }
}

/// Actually quit. Sets `confirmed` FIRST so the ExitRequested hook lets this one through.
fn finish_quit(app: &AppHandle) {
    app.state::<QuitGuard>().confirmed.store(true, Ordering::SeqCst);
    app.exit(0);
}

/// The veto stands and nothing else happens — a cancelled quit is a non-event (no toast).
fn end_prompt(app: &AppHandle) {
    app.state::<QuitGuard>().prompting.store(false, Ordering::SeqCst);
}

/// Every guarded quit path funnels here. `bypass` is ⌥⌘Q — the deliberate, per-quit escape
/// hatch, which is why it is a modifier and not a checkbox inside the dialog.
pub fn request_quit(app: &AppHandle, bypass: bool) {
    let guard = app.state::<QuitGuard>();
    if guard.confirmed.load(Ordering::SeqCst) {
        return; // already on the way out
    }
    if bypass || !guard.ask.load(Ordering::SeqCst) {
        finish_quit(app);
        return;
    }

    let lanes = app.state::<LiveLanes>().snapshot();
    let (busy, idle): (Vec<LiveLane>, Vec<LiveLane>) = lanes.into_iter().partition(|l| is_busy(&l.phase));
    if busy.is_empty() {
        finish_quit(app); // nothing in flight — asking would be a question with nothing behind it
        return;
    }

    // A repeat ⌘Q while the question is outstanding RE-ASKS; it never quits. Repetition
    // reaching the destructive answer is the whole thing this guard is against.
    //
    // Re-asking rather than ignoring matters because the renderer is killed and respawned on
    // its own schedule (the WebContent recycle). If that lands while the dialog is up, the
    // dialog goes with it and `prompting` would otherwise stay set forever — an app that can
    // no longer be quit at all. A second ⌘Q now brings the question back; ⌥⌘Q, which skips
    // this whole path, is the escape hatch if even that fails.
    guard.prompting.store(true, Ordering::SeqCst);
    let round = guard.round.fetch_add(1, Ordering::SeqCst) + 1;
    guard.acked.store(false, Ordering::SeqCst);

    raise_main(app);
    let payload = QuitRequest { lanes: busy.clone(), idle: idle.len() };
    let _ = app.emit("quit:requested", &payload);

    let app = app.clone();
    std::thread::spawn(move || {
        std::thread::sleep(Duration::from_millis(ACK_MS));
        let guard = app.state::<QuitGuard>();
        let stale = guard.round.load(Ordering::SeqCst) != round;
        if stale || guard.acked.load(Ordering::SeqCst) || !guard.prompting.load(Ordering::SeqCst) {
            return; // the webview showed the real dialog, or the question is already answered
        }
        native_fallback(&app, &busy, round);
    });
}

/// The backstop dialog, for when there is no React app to render the real one.
///
/// Known and accepted degradation: `tauri-plugin-dialog` cannot choose which button is
/// default, so Return lands on OK — `Stay open`, which is correct — while Esc maps to Cancel
/// and quits. Mashing Return is the panic reflex and it lands safe. The buttons are NOT
/// inverted to "fix" Esc: that would put the destructive verb on Return, which is worse.
fn native_fallback(app: &AppHandle, busy: &[LiveLane], round: u64) {
    use tauri_plugin_dialog::{DialogExt, MessageDialogButtons};

    let n = busy.len();
    let (message, quit_verb) = if n == 1 {
        (
            "1 agent is still running. Quitting ends it. Its worktree and transcript stay on disk.".to_string(),
            "Quit and end it",
        )
    } else {
        (
            format!("{n} agents are still running. Quitting ends all of them. Their worktrees and transcripts stay on disk."),
            "Quit and end them",
        )
    };
    let stay = app
        .dialog()
        .message(message)
        .title("Quit Operator?")
        .buttons(MessageDialogButtons::OkCancelCustom("Stay open".into(), quit_verb.into()))
        .blocking_show();

    let guard = app.state::<QuitGuard>();
    if guard.round.load(Ordering::SeqCst) != round || !guard.prompting.load(Ordering::SeqCst) {
        return; // the webview woke up and answered while this was open
    }
    if stay {
        end_prompt(app);
    } else {
        finish_quit(app);
    }
}

// --- commands ---------------------------------------------------------------------------

/// The dialog mounted. Cancels the native fallback.
#[tauri::command]
pub fn quit_dialog_shown(guard: State<QuitGuard>) {
    guard.acked.store(true, Ordering::SeqCst);
}

/// The user answered. `quit: false` = Stay open.
#[tauri::command]
pub fn quit_decision(app: AppHandle, quit: bool) {
    if !app.state::<QuitGuard>().prompting.load(Ordering::SeqCst) {
        return; // nothing is being asked — ignore a stray answer
    }
    if quit {
        finish_quit(&app);
    } else {
        end_prompt(&app);
    }
}

/// Mirror of the renderer's preference. See `QuitGuard::ask`.
#[tauri::command]
pub fn quit_set_ask(guard: State<QuitGuard>, ask: bool) {
    guard.ask.store(ask, Ordering::SeqCst);
}

// --- menu -------------------------------------------------------------------------------

/// The app menu, with ONLY the Quit item replaced.
///
/// This starts from `Menu::default` and edits it, rather than composing a menu by hand, and
/// that is load-bearing: Tauri installs the default menu *only* while `Builder::menu` is never
/// called (tauri-2.11.2/src/app.rs:2236-2241), so calling it makes us the owner of the whole
/// menu bar — including the Edit submenu, which is what supplies Copy/Paste/Select All/Undo to
/// the webview on macOS. A hand-rolled menu that forgets it takes paste out of the composer,
/// which would be a worse regression than the one being fixed.
///
/// If the default's shape ever changes under us, we keep the stock menu and log it: ⌘Q goes
/// back to being unguarded, which is today's behaviour, rather than shipping a mangled menu.
pub fn build_menu(app: &AppHandle) -> tauri::Result<Menu<tauri::Wry>> {
    let menu = Menu::default(app)?;

    // macOS app submenu, per tauri-2.11.2/src/menu/menu.rs:190-204:
    // [about, ---, services, ---, hide, hide_others, ---, quit] — quit last, 8 items.
    const APP_MENU_LEN: usize = 8;

    let Some(MenuItemKind::Submenu(app_menu)) = menu.items()?.into_iter().next() else {
        eprintln!("[quit] default menu has no app submenu — leaving it stock, ⌘Q unguarded");
        return Ok(menu);
    };
    let items = app_menu.items()?;
    if items.len() != APP_MENU_LEN || !matches!(items.last(), Some(MenuItemKind::Predefined(_))) {
        eprintln!("[quit] app submenu is not the expected shape — leaving it stock, ⌘Q unguarded");
        return Ok(menu);
    }
    if let Err(e) = app_menu.remove_at(APP_MENU_LEN - 1) {
        eprintln!("[quit] could not remove the predefined Quit item ({e}) — leaving it stock, ⌘Q unguarded");
        return Ok(menu);
    }
    // NOTHING below returns Err. Past this point the stock Quit item is already gone, and a
    // `?` here would abort `Builder::build` — trading a missing Quit item for no menu bar at
    // all, which is to say no Copy/Paste either. Worst case we log and ship a Quit-less app
    // menu; ⌘Q then does nothing rather than quitting unguarded.
    //
    // The second item is the per-quit escape hatch, spelled out so it is learnable rather than
    // folklore. muda 0.19.2 has no alternate-item support (no `alternate` anywhere in the
    // crate), so it is a plain visible row rather than the ⌥-held reveal the design asked for.
    for (id, text, accel) in [
        (MENU_QUIT, "Quit Operator", "CmdOrCtrl+Q"),
        (MENU_QUIT_NOW, "Quit Without Asking", "Alt+CmdOrCtrl+Q"),
    ] {
        match MenuItem::with_id(app, id, text, true, Some(accel)) {
            Ok(item) => {
                if let Err(e) = app_menu.append(&item) {
                    eprintln!("[quit] could not append {id} to the app menu: {e}");
                }
            }
            Err(e) => eprintln!("[quit] could not build the {id} menu item: {e}"),
        }
    }
    Ok(menu)
}

/// Route the two app-menu items. Anything else is left to whoever else is listening.
pub fn on_menu_event(app: &AppHandle, id: &str) {
    match id {
        MENU_QUIT => request_quit(app, false),
        MENU_QUIT_NOW => request_quit(app, true),
        _ => {}
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    /// The one decision in this module that is pure, and the one the design turns on:
    /// `waiting` is busy, `idle` is not.
    #[test]
    fn busy_means_mid_turn_or_blocked_on_you_but_never_idle() {
        assert!(is_busy("running"));
        assert!(is_busy("compacting"));
        assert!(is_busy("waiting"));
        assert!(!is_busy("idle"));
        assert!(!is_busy(""));
    }

    #[test]
    fn the_guard_starts_armed() {
        let g = QuitGuard::default();
        assert!(g.ask.load(Ordering::SeqCst), "asking must be the default, not opt-in");
        assert!(!g.confirmed.load(Ordering::SeqCst));
        assert!(!g.prompting.load(Ordering::SeqCst));
    }
}

File: electron/src/main/quit.ts
// The quit guard: ask before quitting while agents are still working.
//
// Ported from `src-tauri/src/quit.rs`, and CHEAPER here — `event.preventDefault()` on
// `before-quit` IS the veto that Tauri had to build around `RunEvent::ExitRequested`.
//
// RUST OWNED THE VETO AND THE LANE COUNT, and main owns them here for the same reason: the
// accident this guards left the webview navigated away with no React app at all, so a count
// read from a frontend store is absent in exactly the case it is needed.
import { app, BrowserWindow, dialog } from 'electron'

export interface QuitLane { terminalId: string; project: string; phase: string }

/** The phases that mean "mid-turn", and the one decision in this module that is pure.
 *
 *  `waiting` IS BUSY, deliberately: an agent blocked on YOU is the precise lane you forgot
 *  about. `idle` is not, because a guard that fires on nearly every quit is one that trains its
 *  own dismissal.
 *
 *  Ported from `is_busy` in quit.rs. The Electron wiring originally asked for
 *  `running || compacting` and dropped `waiting` — which is to say it stayed silent about
 *  exactly the lane the guard exists for. */
export function isBusy(phase: string): boolean {
  return phase === 'running' || phase === 'compacting' || phase === 'waiting'
}

/** How long to wait for the renderer to confirm it mounted the dialog before falling back to a
 *  native one. The renderer may be mid-reload, crashed, or navigated away — which is precisely
 *  the accident this exists for — and a guard that silently does nothing in that case is not a
 *  guard. */
const DIALOG_ACK_MS = 400

export class QuitGuard {
  private ask = true
  private quitting = false
  /** THE LATCH. One answer per question — see `decide`. */
  private answered = false
  /** Which dialog currently owns the open question, so a second ⌘Q cannot raise another one on
   *  top of it. `nativeAsk` is `await`ed on a modal sheet, and a second sheet queued behind the
   *  first is a question the user answers twice. */
  private asking: 'none' | 'renderer' | 'native' = 'none'
  private ackTimer: NodeJS.Timeout | null = null

  constructor(
    private readonly getWindow: () => BrowserWindow | null,
    private readonly busyLanes: () => QuitLane[],
    private readonly idleCount: () => number,
    private readonly onTeardown: () => void,
  ) {}

  /** Mirror of the "Ask before quitting…" switch, which lives in localStorage with the app's
   *  other prefs. Main cannot read that, and must still be right when the renderer is gone. */
  setAsk(ask: boolean): void { this.ask = ask }

  /** The window a modal must hang on, RAISED first — and the reason every `showMessageBox` in
   *  this file takes one.
   *
   *  THE FREEZE. `dialog.showMessageBox(options)` with no parent window is Electron's
   *  application-modal branch: on macOS it runs `[NSAlert runModal]`, a nested run loop on the
   *  browser process's UI thread. Every window stops painting and the app stops answering — the
   *  hang the user reported as Operator freezing on quit. Passing a window instead takes the
   *  `beginSheetModalForWindow:` branch, which is document-modal: the sheet drops out of the
   *  title bar, the run loop keeps turning, and the promise resolves on the callback.
   *
   *  RAISED, not merely passed, and both halves matter. A sheet on a minimized or hidden window
   *  is a question nobody can see attached to an app that will not quit, which is the freeze
   *  again from the user's side. It also gives the second fix its teeth: a document-modal sheet
   *  blocks the web contents underneath, so a stale renderer dialog left over from a late ack
   *  cannot be clicked while the sheet is up.
   *
   *  `undefined` when there is no window at all — the app-modal call is still the right
   *  fallback there, because there is nothing to attach to and the question must still be
   *  asked. */
  private modalParent(): BrowserWindow | undefined {
    const win = this.getWindow()
    if (!win || win.isDestroyed()) return undefined
    // A window can be torn down between the check and any of these; a guard that throws here
    // is a quit that hangs, which is the thing being fixed.
    try {
      if (win.isMinimized()) win.restore()
      if (!win.isVisible()) win.show()
      win.focus()
    } catch { /* gone from under us — fall through and ask anyway */ }
    return win
  }

  /** THE UPDATE'S QUESTION, asked ONCE and by the guard itself.
   *
   *  An update used to reach `quitAndInstall` without the guard knowing, so the guard asked its
   *  own question on the way out and its veto CANCELLED the install. The user pressed "Install &
   *  Restart" and stayed on the old version. So the guard asks first, in its own voice, with the
   *  same busy-lane count it would have used — and `beginQuitting()` below then makes sure it
   *  cannot ask a second time.
   *
   *  NATIVE, not the renderer's dialog, and deliberately: the renderer dialog exists for styling,
   *  and the failure being fixed here is a prompt that got lost. An install prompt that a
   *  mid-reload renderer can swallow would be the same bug with a nicer font. */
  async askInstall(version: string): Promise<boolean> {
    const lanes = this.busyLanes()
    const detail = lanes.length
      ? `${lanes.map((l) => `${l.project} (${l.phase})`).join(', ')}\n\nRestarting ends their terminals. The update is already downloaded — if you'd rather not now, it installs the next time you quit.`
      : 'The update is already downloaded. Operator will restart once it is installed.'
    const opts = {
      type: 'question' as const,
      buttons: ['Not now', 'Install and restart'],
      defaultId: 1,
      cancelId: 0,
      message: lanes.length
        ? `Install ${version} and restart? ${lanes.length} lane${lanes.length === 1 ? ' is' : 's are'} busy`
        : `Install ${version} and restart?`,
      detail,
    }
    // Parented for the same reason as `nativeAsk` — this one is asked while the app is very
    // much alive and mid-update, so freezing the UI thread here is if anything worse.
    const parent = this.modalParent()
    const { response } = parent
      ? await dialog.showMessageBox(parent, opts)
      : await dialog.showMessageBox(opts)
    return response === 1
  }

  /** DISARM THE VETO. Everything after this point treats the quit as already decided, which is
   *  what stops `before-quit` asking a question the user has just answered — and cancelling the
   *  installer's own quit with it. */
  beginQuitting(): void { this.quitting = true }

  install(): void {
    app.on('before-quit', (e) => {
      if (this.quitting || !this.ask) { this.onTeardown(); return }
      const lanes = this.busyLanes()
      if (!lanes.length) { this.onTeardown(); return }

      e.preventDefault()
      const win = this.getWindow()
      const payload = { lanes, idle: this.idleCount() }

      // A SECOND ⌘Q WHILE A QUESTION IS OPEN. Re-emitting to the renderer is deliberate — it is
      // how a renderer that respawned mid-question gets its dialog back — but nothing else may
      // restart. Re-arming the timer could raise a native sheet over a renderer dialog that is
      // already up, and clearing the latch would let an answer to the FIRST dialog land after
      // an answer to the second.
      if (this.asking !== 'none') {
        if (this.asking === 'renderer' && win && !win.isDestroyed()) {
          win.webContents.send('operator-event:onQuitRequested', payload)
        }
        return
      }

      this.answered = false
      if (win && !win.isDestroyed()) {
        this.asking = 'renderer'
        win.webContents.send('operator-event:onQuitRequested', payload)
        // If the renderer never acks, it is not in a state to ask anything — go native.
        this.ackTimer = setTimeout(() => {
          this.ackTimer = null
          if (this.answered) return
          void this.nativeAsk(lanes)
        }, DIALOG_ACK_MS)
      } else {
        void this.nativeAsk(lanes)
      }
    })
  }

  /** The renderer mounted its dialog — cancel the native fallback.
   *
   *  A LATE ACK IS NOT AN ERROR AND IS NOT OBEYED. If it arrives after `DIALOG_ACK_MS` the
   *  fallback sheet is already up, and the renderer has already shown its own dialog by the
   *  time it calls this (it acks in the same turn it sets the state). Both are then on screen.
   *  The sheet is document-modal on that window, so it covers the renderer's dialog and takes
   *  the click; whichever is answered, `decide`'s latch keeps the first answer. */
  dialogShown(): void {
    if (this.ackTimer) { clearTimeout(this.ackTimer); this.ackTimer = null }
  }

  /** `false` = Stay open. The veto then stands and nothing else happens.
   *
   *  ONE ANSWER PER QUESTION, and the latch is the fix for a real disagreement rather than
   *  tidiness. A renderer that acks after `DIALOG_ACK_MS` leaves TWO dialogs up, and they can
   *  be answered differently — the old code took both: native "Stay open" left the veto
   *  standing and did not close the question, so a later "Quit anyway" from the stale renderer
   *  dialog quit an app the user had just chosen to keep. The first answer is the user's
   *  answer; the second is a stale dialog's echo and is dropped. */
  decide(quit: boolean): void {
    if (this.answered) return
    this.answered = true
    this.asking = 'none'
    if (this.ackTimer) { clearTimeout(this.ackTimer); this.ackTimer = null }
    if (!quit) return
    this.quitting = true
    this.onTeardown()
    app.quit()
  }

  private async nativeAsk(lanes: QuitLane[]): Promise<void> {
    if (this.answered) return
    this.asking = 'native'
    const names = lanes.map((l) => `${l.project} (${l.phase})`).join(', ')
    const opts = {
      type: 'warning' as const,
      buttons: ['Stay open', 'Quit anyway'],
      defaultId: 0,
      cancelId: 0,
      message: `${lanes.length} agent${lanes.length === 1 ? ' is' : 's are'} still working`,
      detail: `${names}\n\nQuitting ends their terminals.`,
    }
    const parent = this.modalParent()
    const { response } = parent
      ? await dialog.showMessageBox(parent, opts)
      : await dialog.showMessageBox(opts)
    // BOTH answers are recorded, "Stay open" included. It used to fall off the end, leaving the
    // question open — which is precisely what let a stale renderer dialog quit afterwards.
    this.decide(response === 1)
  }
}
